### PR TITLE
[cuda] Port over CUDA stream-based command buffer impl

### PR DESCRIPTION
**Pull request by antiagainst**
This commit uses existing CUDA HAL driver's stream command buffer
implementation. Improvements include removing context wrapper and
adding various Tracy tracking markers. Various places are cleaned
up, e.g., barrier/event impl. Inline execution support is dropped
due to its incompatibility with semaphores and recorded command
buffers. With it now we can enable CUDA stream-based tests.

Progress towards https://github.com/openxla/iree/issues/13245

### CODE REVIEW

**Comment by antiagainst** [comment]
Note that the first and second commit is just copying over existing code and mass renaming. Only the third one needs review.

File: experimental/cuda2/cuda_device.c [hunk 1]
```diff
@@ -194,6 +199,18 @@ static iree_status_t iree_hal_cuda2_device_create_internal(
         host_allocator, &device->device_allocator);
   }
 
+  if (iree_status_is_ok(status) &&
+      params->command_buffer_mode == IREE_HAL_CUDA_COMMAND_BUFFER_MODE_STREAM) {
+    status = iree_hal_cuda2_stream_command_buffer_create(
```
Comment (benvanik) [inline experimental/cuda2/cuda_device.c]: I can't figure out what this is used for, but it shouldn't exist - we can't reuse command buffers like this or have global state - multiple pieces of code may be using the same device and submitting work that overlaps - just make new command buffers as you need them

**Review by benvanik — CHANGES_REQUESTED** [review]

File: experimental/cuda2/cuda_device.c (hunk 1 above)
Comment (antiagainst) [reply experimental/cuda2/cuda_device.c]: Ah, right, good catch! Missed this part when porting over existing code.. Changed to do the allocation where it's used.

**Comment by benvanik** [comment]
ahhh I was confused - it was called `deferred_command_buffer` in the port but was really a stream command buffer - it was `stream_command_buffer` in the original so I thought you were changing it to a deferred command buffer for some reason - it's best not to share anyway, though, so still a good fix!

**Review by benvanik — APPROVED** [review]

**Comment by antiagainst** [comment]
Yeah, I was thinking that instead of naming it following its type, naming it to reflect how it's used may be better.. Naming is hard. ;P Thanks for the comments again!

**Comment by antiagainst** [comment]
Oh shoot. First time merging a pull request using the GitHub app. I chose to do squash and merge, somehow it landed asa rebase and merge. Didn't know what I messed up with the UI. :( 

The commits are somewhat reasonably separated and messaged though. Hopefully not messing up with the history too much..

**Pull request merged**